src/Conversion/ONNXToKrnl/ML/CategoryMapper.cpp

```diff
@@ -0,0 +1,294 @@
+                  "default_string", default_string)
+            : nullptr;
+
+    // Lookup the index in the perfect hash table corresponding to
+    // each input value.
+    BuildKrnlLoop inputLoops(rewriter, loc, rank);
+    inputLoops.createDefineAndIterateOp(X);
+    rewriter.setInsertionPointToStart(inputLoops.getIterateBlock());
+    {
+      // Get loop indices.
+      SmallVector<IndexExpr, 4> IVs;
+      for (decltype(rank) i = 0; i < rank; ++i) {
+        Value iv = inputLoops.getInductionVar(i);
+        IVs.emplace_back(DimIndexExpr(iv));
+      }
+
+      // Determine the index of 'inputElem' in the perfect hash table
+      // 'pHash'. Note: the index might not be valid (this happens
+      // when the 'inputElem' is not present in the perfect hash
+      // table).
+      Value inputElem = createKrnl.loadIE(X, IVs);
+      Value index, isIndexValid;
+      std::tie(index, isIndexValid) =
+          findIndex(inputElem, elementType, perfectHashTable,
+              constantForCatsInt64s, constantForCatsStrings, createKrnl);
+
+      // Store the final result.
+      scf::IfOp ifOp = rewriter.create<scf::IfOp>(
+          loc, isIndexValid, /*withElseRegion=*/true);
+      storeResult(index, elementType, ifOp, constantForCatsInt64s,
+          constantForCatsStrings, defaultInt64, defaultString, alloc, IVs,
+          createKrnl, rewriter);
+    }
+
+    rewriter.replaceOp(op, alloc);
+
+    LLVM_DEBUG({
+      FuncOp function = getContainingFunction(op);
+      assert(function && "Could not find parent function");
+      dbgs() << "function: " << function << "\n";
+    });
+
+    return success();
+  }
+
+private:
+  Attribute getElemAttr(ArrayAttr arr, int32_t idx) const {
+    return arr.getValue()[idx];
+  }
+
+  // Generate a perfect hash table for the input dictionary.
+  // Depending on the runtime type 'elementType' (the type of the element of
+  // the input tensor) this function created a perfect hash table for:
+  //  - cats_int64s if elementType is a int64_t
+  //  - cats_strings if elementType is a string
+  PerfectHashTable createPerfectHashTable(DenseElementsAttr cats_int64s,
+      DenseElementsAttr cats_strings, ArrayAttr cats_int64s_ArrayAttr,
+      ArrayAttr cats_strings_ArrayAttr, Type elementType,
+      const KrnlBuilder &createKrnl) const {
+    OpBuilder builder = createKrnl.getBuilder();
+    PerfectHashTable res;
+
+    // Create constants to hold the arrays 'G' and 'V'.
+    auto createConstants = [&](const SmallVector<int32_t> &G,
+                               const SmallVector<int32_t> &V) {
+      assert(V.size() == G.size() && "V and G should have the same size");
+
+      MemRefType type = MemRefType::get(
+          {static_cast<int64_t>(V.size())}, builder.getIntegerType(32));
+      res.G = createKrnl.constant(type, "G", builder.getI32VectorAttr(G));
+      res.V = createKrnl.constant(type, "V", builder.getI32VectorAttr(V));
```

Comment:
@tungld goot catch. Will fix.

Reply:
> I have a general question: can we use int64_t instead of int32_t for G and V?CategoryMapper's input type is int64_t

Since the map size can be large, it would make sense to me to have the smaller data type for the one we create (G if I recall correctly)